### PR TITLE
Add multiple REX options. See issues #204 #210

### PR DESCRIPTION
**Pull request by alosadagrande**
Add multiple options to remote execution in order to solve enhancement #204 and #210:

--rex-capsules: Comma separated list of capsules to install Foreman's SSH keys for remote execution. It will iterate all over the list of capsules. If one fails user is notified by the task keeps going until all capsules were tried.

--rex-urlkeyfile: HTTP/S location to install a file containing one or multiple Foreman's SSH keys for remote execution. Instead of iterate you can create file on /pub folder  that contains all ssh public keys of several capsules and tell bootstrap.py to fech it and install on the local user authorized keys file.

--rex-authpath. Local folder where is placed the authorized_keys file to install Foreman's SSH keys for remote execution. Default ~/.ssh. I found that several customers specifies AuthorizedKeysFile on sshd_config on a different path.

### CODE REVIEW

File: bootstrap.py [hunk 1]
```diff
@@ -880,6 +889,9 @@ def exec_service(service, command, failonerror=True):
     parser.add_option("--unmanaged", dest="unmanaged", action="store_true", help="Add the server as unmanaged. Useful to skip provisioning dependencies.")
     parser.add_option("--rex", dest="remote_exec", action="store_true", help="Install Foreman's SSH key for remote execution.", default=False)
     parser.add_option("--rex-user", dest="remote_exec_user", default="root", help="Local user used by Foreman's remote execution feature.")
+    parser.add_option("--rex-capsules", dest="remote_exec_capsules", default="", help="Comma separated list of capsules to install Foreman's SSH keys for remote execution.")
```
Comment (evgeni) [inline bootstrap.py]: If there is no default, you can omit the `default` key completely.

File: bootstrap.py [hunk 2]
```diff
@@ -449,12 +458,12 @@ def install_foreman_ssh_key():
         return
     if os.path.isfile(foreman_ssh_authfile):
         if foreman_ssh_key in open(foreman_ssh_authfile, 'r').read():
-            print_generic("Foreman's SSH key is already present in %s" % foreman_ssh_authfile)
+            print_generic("Foreman's SSH key/s already present in %s" % foreman_ssh_authfile)
```
Comment (evgeni) [inline bootstrap.py]: why this change? do you want the sentence to be "ssh key (or keys) already present"? given the function still adds one key at a time, I think we should not change that. But we could add the remote-url to indicate which key was present?

File: bootstrap.py [hunk 3]
```diff
@@ -449,12 +458,12 @@ def install_foreman_ssh_key():
...
             return
     output = os.fdopen(os.open(foreman_ssh_authfile, os.O_WRONLY | os.O_CREAT, 0600), 'a')
     output.write(foreman_ssh_key)
     os.chown(foreman_ssh_authfile, userpw.pw_uid, userpw.pw_gid)
-    print_generic("Foreman's SSH key was added to %s" % foreman_ssh_authfile)
+    print_generic("Foreman's SSH key/s added to %s" % foreman_ssh_authfile)
```
Comment (evgeni) [inline bootstrap.py]: same as above with s/present/added/ ;)

File: bootstrap.py [hunk 4]
```diff
@@ -426,19 +426,28 @@ def fully_update_the_box():
...
+            return
+    else:
+        foreman_ssh_dir = os.sep.join([userpw.pw_dir, '.ssh'])
+        foreman_ssh_authfile = os.sep.join([foreman_ssh_dir, 'authorized_keys'])
     if not os.path.isdir(foreman_ssh_dir):
         os.mkdir(foreman_ssh_dir, 0700)
         os.chown(foreman_ssh_dir, userpw.pw_uid, userpw.pw_gid)
     try:
-        foreman_ssh_key = urllib2.urlopen("https://%s:9090/ssh/pubkey" % options.foreman_fqdn).read()
+        foreman_ssh_key = urllib2.urlopen("%s" % remote_url).read()
```
Comment (evgeni) [inline bootstrap.py]: `foreman_ssh_key = urllib2.urlopen(remote_url).read()`

no need for string format

File: bootstrap.py [hunk 5]
```diff
@@ -426,19 +426,28 @@ def fully_update_the_box():
...
-    foreman_ssh_authfile = os.sep.join([foreman_ssh_dir, 'authorized_keys'])
+    if options.remote_exec_authpath:
+        if os.path.isdir(options.remote_exec_authpath):
+            foreman_ssh_dir = options.remote_exec_authpath
+            foreman_ssh_authfile = options.remote_exec_authpath + "/authorized_keys"
+        else:
+            print_error("Foreman's SSH keys not installed. Directory where authorized_keys file must be located is not found: %s" % options.remote_exec_authpath)
+            return
+    else:
+        foreman_ssh_dir = os.sep.join([userpw.pw_dir, '.ssh'])
```
Comment (evgeni) [inline bootstrap.py]: can we make the default value of `options.remote_exec_authpath` to be `os.path.join(userpw.pw_dir, '.ssh')` then we could skip this if/else block.

note that I changed `os.sep.join([userpw.pw_dir, '.ssh'])` to `os.path.join(userpw.pw_dir, '.ssh')` which I think is more correct here.

File: bootstrap.py [hunk 6]
```diff
@@ -426,19 +426,28 @@ def fully_update_the_box():
...
+    authorized keys file location, so that remote execution becomes possible.
+    By default is ~/.ssh/authorized_keys
     """
     userpw = pwd.getpwnam(options.remote_exec_user)
-    foreman_ssh_dir = os.sep.join([userpw.pw_dir, '.ssh'])
-    foreman_ssh_authfile = os.sep.join([foreman_ssh_dir, 'authorized_keys'])
+    if options.remote_exec_authpath:
+        if os.path.isdir(options.remote_exec_authpath):
+            foreman_ssh_dir = options.remote_exec_authpath
+            foreman_ssh_authfile = options.remote_exec_authpath + "/authorized_keys"
```
Comment (evgeni) [inline bootstrap.py]: There is no requirement for the file to be called `authorized_keys`. In fact, some of my boxes have `AuthorizedKeysFile /etc/ssh/authorized_keys/%u` (with `%u` being the username). So I think we just should take the full path at the command line and not alter it at all.

**Review by evgeni — CHANGES_REQUESTED** [review]

File: bootstrap.py (hunk 6 above)
Comment (alosadagrande) [reply bootstrap.py]: OK

File: bootstrap.py (hunk 5 above)
Comment (alosadagrande) [reply bootstrap.py]: Actually from the first comment options.remote_exec_authpath is the complete path to authorizedKey file not a dir.

File: bootstrap.py (hunk 4 above)
Comment (alosadagrande) [reply bootstrap.py]: ok

File: bootstrap.py (hunk 2 above)
Comment (alosadagrande) [reply bootstrap.py]: OK. Depending on the rex option you're adding one key at a time or multiple when copying a remote file containing multiples keys. So if adding multiples you're checking multiple keys at a time.

File: bootstrap.py (hunk 1 above)
Comment (alosadagrande) [reply bootstrap.py]: OK

File: bootstrap.py (hunk 3 above)
Comment (alosadagrande) [reply bootstrap.py]: ok. same

File: bootstrap.py [hunk 7]
```diff
@@ -426,19 +426,24 @@ def fully_update_the_box():
...
-        os.chown(foreman_ssh_dir, userpw.pw_uid, userpw.pw_gid)
+    if not options.remote_exec_authpath:
+        userpw = pwd.getpwnam(options.remote_exec_user)
+        options.remote_exec_authpath = os.path.join(*[userpw.pw_dir, '.ssh', 'authorized_keys'])
+        foreman_ssh_dir = os.path.join(*[userpw.pw_dir, '.ssh'])
+        if not os.path.isdir(foreman_ssh_dir):
+            os.mkdir(foreman_ssh_dir, 0700)
+            os.chown(foreman_ssh_dir, userpw.pw_uid, userpw.pw_gid)
+    elif not os.path.isfile(options.remote_exec_authpath):
+        print_error("Foreman's SSH key not installed. File where authorized_keys must be located is not found: %s" % options.remote_exec_authpath)
```
Comment (evgeni) [inline bootstrap.py:531]: Why? We can just create the file (and we actually do)?

File: bootstrap.py [hunk 8]
```diff
@@ -880,6 +885,9 @@ def exec_service(service, command, failonerror=True):
     parser.add_option("--unmanaged", dest="unmanaged", action="store_true", help="Add the server as unmanaged. Useful to skip provisioning dependencies.")
     parser.add_option("--rex", dest="remote_exec", action="store_true", help="Install Foreman's SSH key for remote execution.", default=False)
     parser.add_option("--rex-user", dest="remote_exec_user", default="root", help="Local user used by Foreman's remote execution feature.")
+    parser.add_option("--rex-capsules", dest="remote_exec_capsules", help="Comma separated list of capsules to install Foreman's SSH keys for remote execution.")
+    parser.add_option("--rex-urlkeyfile", dest="remote_exec_url", help="HTTP/S location to install a file containing one or multiple Foreman's SSH keys for remote execution.")
+    parser.add_option("--rex-authpath", dest="remote_exec_authpath", help="Local folder where is placed the authorized_keys file to install Foreman's SSH keys for remote execution. Default ~/.ssh")
```
Comment (evgeni) [inline bootstrap.py]: Local file, not folder :)

File: bootstrap.py [hunk 9]
```diff
@@ -426,19 +426,24 @@ def fully_update_the_box():
...
     """
-    userpw = pwd.getpwnam(options.remote_exec_user)
-    foreman_ssh_dir = os.sep.join([userpw.pw_dir, '.ssh'])
-    foreman_ssh_authfile = os.sep.join([foreman_ssh_dir, 'authorized_keys'])
-    if not os.path.isdir(foreman_ssh_dir):
-        os.mkdir(foreman_ssh_dir, 0700)
-        os.chown(foreman_ssh_dir, userpw.pw_uid, userpw.pw_gid)
+    if not options.remote_exec_authpath:
+        userpw = pwd.getpwnam(options.remote_exec_user)
+        options.remote_exec_authpath = os.path.join(*[userpw.pw_dir, '.ssh', 'authorized_keys'])
```
Comment (evgeni) [inline bootstrap.py]: `os.path.join(userpw.pw_dir, '.ssh', 'authorized_keys')` should do it too, no?

File: bootstrap.py [hunk 10]
```diff
@@ -426,19 +426,24 @@ def fully_update_the_box():
...
+        foreman_ssh_dir = os.path.join(*[userpw.pw_dir, '.ssh'])
```
Comment (evgeni) [inline bootstrap.py]: same as above

**Review by evgeni — DISMISSED** [review]

File: bootstrap.py (hunk 7 above)
Comment (alosadagrande) [reply bootstrap.py:531]: That's the case when the path is entered by user. I think must be warned.

If there is no remote_exec_autpath we pick the default one (~/.ssh/authorized_keys) where we create the folder if needed

Comment (alosadagrande) [reply bootstrap.py:531]: We should not create the complete path to the remote_exec_autpath if it's entered by user, otherwise we'll probably face security issues in the future.

**Comment by evgeni** [comment]
rebased and squashed this (thanks @fcami for the rebase work)

**Comment by evgeni** [comment]
@sideangleside as I got my hands dirty in here, you'll have to do the final review

**Comment by beav** [comment]
@alosadagrande it looks like this needs a rebase

**Comment by fcami** [comment]
@beav @alosadagrande @evgeni I'll rebase

**Comment by fcami** [comment]
@beav @evgeni could you please review and pull from https://github.com/fcami/katello-client-bootstrap/tree/multi-key-rex - it seems OK to me but I haven't tested it yet.

**Comment by beav** [comment]
@fcami the change looks OK to me but I didn't test it either.

If @alosadagrande doesn't update this PR, the best bet is for you to create a new branch, then do "git cherry-pick -x 299b0349c3168e0cf0a5473e2a1a3e967b98a0ef" and resolve the conflict, then do "git add bootstrap.py" and "git commit". That way it will keep the authorship information correct and you can then create a new PR and we can close this one out.

**Comment by fcami** [comment]
@beav thanks - please see https://github.com/Katello/katello-client-bootstrap/pull/253 then

**Comment by evgeni** [comment]
obsoleted by https://github.com/Katello/katello-client-bootstrap/pull/281
